modules/gdscript/gdscript_analyzer.cpp

```diff
@@ -2089,35 +2175,50 @@ void GDScriptAnalyzer::resolve_assert(GDScriptParser::AssertNode *p_assert) {
 void GDScriptAnalyzer::resolve_match(GDScriptParser::MatchNode *p_match) {
 	reduce_expression(p_match->test);
 
+	Vector<GDScriptParser::IdentifierNode *> identifiers = deduce_nullable_narrowing(p_match->test);
+	push_narrowed_nullables(identifiers);
 	for (int i = 0; i < p_match->branches.size(); i++) {
-		resolve_match_branch(p_match->branches[i], p_match->test);
+		resolve_match_branch(p_match->branches[i], p_match->test, identifiers);
 
 		decide_suite_type(p_match, p_match->branches[i]);
 	}
+	pop_narrowed_nullables(identifiers);
 }
 
-void GDScriptAnalyzer::resolve_match_branch(GDScriptParser::MatchBranchNode *p_match_branch, GDScriptParser::ExpressionNode *p_match_test) {
+void GDScriptAnalyzer::resolve_match_branch(GDScriptParser::MatchBranchNode *p_match_branch, GDScriptParser::ExpressionNode *p_match_test, Vector<GDScriptParser::IdentifierNode *> p_nullable_identifier) {
 	for (int i = 0; i < p_match_branch->patterns.size(); i++) {
-		resolve_match_pattern(p_match_branch->patterns[i], p_match_test);
+		bool is_nullable = false;
+		resolve_match_pattern(p_match_branch->patterns[i], p_match_test, &is_nullable);
+		if (is_nullable) {
+			pop_narrowed_nullables(p_nullable_identifier);
+		}
 	}
 
 	resolve_suite(p_match_branch->block);
 
 	decide_suite_type(p_match_branch, p_match_branch->block);
+
+	// Re-push our nullable identifiers so the next branch has access to it
+	push_narrowed_nullables(p_nullable_identifier);
 }
 
-void GDScriptAnalyzer::resolve_match_pattern(GDScriptParser::PatternNode *p_match_pattern, GDScriptParser::ExpressionNode *p_match_test) {
+void GDScriptAnalyzer::resolve_match_pattern(GDScriptParser::PatternNode *p_match_pattern, GDScriptParser::ExpressionNode *p_match_test, bool *p_is_nullable) {
 	if (p_match_pattern == nullptr) {
 		return;
 	}
-
+#define SET_NULLABLE(value)         \
```

Comment:
I know Godot's style guide stance on lambdas but if ever there was a case for one...